File: scala/friesian/src/main/java/com/intel/analytics/bigdl/friesian/example/SimilarityClient.java
/*
 * Copyright 2021 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.friesian.example;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;
import com.google.protobuf.Empty;
import com.intel.analytics.bigdl.friesian.serving.grpc.generated.recall.RecallGrpc;
import com.intel.analytics.bigdl.friesian.serving.grpc.generated.recall.RecallProto.Candidates;
import com.intel.analytics.bigdl.friesian.serving.grpc.generated.recall.RecallProto.Query;
import com.intel.analytics.bigdl.friesian.serving.grpc.generated.recall.RecallProto.ServerMessage;
import com.intel.analytics.bigdl.friesian.serving.utils.CMDParser;
import com.intel.analytics.bigdl.friesian.serving.utils.Utils;
import com.intel.analytics.bigdl.grpc.JacksonJsonSerializer;
import io.grpc.Channel;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;
import com.intel.analytics.bigdl.friesian.serving.utils.TimerMetrics;
import com.intel.analytics.bigdl.friesian.serving.utils.TimerMetrics$;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.core.config.Configurator;

import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class SimilarityClient {
    private static final Logger logger = LogManager.getLogger(com.intel.analytics.bigdl.friesian.example.SimilarityClient.class.getName());

    private final RecallGrpc.RecallBlockingStub blockingStub;

    private static MetricRegistry metrics = new MetricRegistry();
    private static Timer searchTimer = metrics.timer("friesian.similarity");

    public SimilarityClient(Channel channel) {
        blockingStub = RecallGrpc.newBlockingStub(channel);
    }

    public void search(int id, int k) {

        Query request = Query.newBuilder().setUserID(id).setK(k).build();

        Candidates candidates;
        try {
            candidates = blockingStub.searchCandidates(request);
            printCandidates(id, candidates);
        } catch (StatusRuntimeException e) {
            logger.warn("RPC failed: " + e.getStatus().toString());
        }
    }

    public void getMetrics() {
        Empty request = Empty.newBuilder().build();

        ServerMessage msg;
        try {
            msg = blockingStub.getMetrics(request);
        } catch (StatusRuntimeException e) {
            logger.warn("RPC failed: " + e.getStatus().toString());
            return;
        }

        logger.info("Got metrics: " + msg.getStr());
    }

    public void resetMetrics() {
        Empty request = Empty.newBuilder().build();
        try {
            blockingStub.resetMetrics(request);
        } catch (StatusRuntimeException e) {
            logger.warn("RPC failed: " + e.getStatus().toString());
        }
    }

    public void getClientMetrics() {
        JacksonJsonSerializer jacksonJsonSerializer = new JacksonJsonSerializer();
        Set<String> keys = metrics.getTimers().keySet();
        List<TimerMetrics> timerMetrics = keys.stream()
                .map(key ->
                        TimerMetrics$.MODULE$.apply(key, metrics.getTimers().get(key)))
                .collect(Collectors.toList());
        String jsonStr = jacksonJsonSerializer.serialize(timerMetrics);
        logger.info("Client metrics: " + jsonStr);
    }

    public void resetClientMetrics() {
        metrics = new MetricRegistry();
        searchTimer = metrics.timer("recall.search");
    }

    public void printCandidates(int id, Candidates candidates) {
        List<Integer> candidateList = candidates.getCandidateList();
        System.out.printf("ID %d: Candidates: ", id);
        StringBuilder sb = new StringBuilder();
        for (Integer candidate: candidateList) {
            sb.append(candidate).append('\t');
        }
        sb.append('\n');
        System.out.print(sb);
    }

    /** Issues several different requests and then exits. */
    public static void main(String[] args) throws InterruptedException {
        Configurator.setLevel("org", Level.ERROR);

        CMDParser cmdParser = new CMDParser();
        cmdParser.addOption("target", "The server to connect to.", "localhost:8085");
        cmdParser.addOption("dataDir", "The data file.", "item_ebd.parquet");

        cmdParser.parseOptions(args);
        String target = cmdParser.getOptionValue("target");
        String dir = cmdParser.getOptionValue("dataDir");

        ManagedChannel channel = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
        com.intel.analytics.bigdl.friesian.example.SimilarityClient client = new com.intel.analytics.bigdl.friesian.example.SimilarityClient(channel);

        int[] item_list = Utils.loadUserData(dir, "tweet_id", 100);

        for (int id: item_list) {
            Timer.Context searchContext = searchTimer.time();
            client.search(id, 10);
            searchContext.stop();
        }
        client.getClientMetrics();
        client.getMetrics();
        client.resetMetrics();
        channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
    }
}

File: python/friesian/example/serving/similarity_client.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import logging
import grpc
from friesian.example.serving.generated import recall_pb2_grpc, recall_pb2
import os
import pandas as pd
import argparse
import time
import threading


if 'http_proxy' in os.environ:
    del os.environ['http_proxy']
if 'https_proxy' in os.environ:
    del os.environ['https_proxy']


class Timer:
    def __init__(self):
        self.count = 0
        self.total = 0

    def reset(self):
        self.count = 0
        self.total = 0

    def add(self, time_elapse):
        self.total += time_elapse * 1000
        self.count += 1

    def get_stat(self):
        return self.count, self.total / self.count


class SimilarityClient():
    def __init__(self, stub):
        self.stub = stub

    def search(self, id, k):
        request = recall_pb2.Query(userID=id, k=k)
        try:
            candidates = self.stub.searchCandidates(request)
            return candidates.candidate
        except Exception as e:
            logging.warning("RPC failed:{}".format(e))
            return

    def getMetrics(self):
        try:
            msg = self.stub.getMetrics(recall_pb2.ServerMessage())
        except Exception as e:
            logging.warning("RPC failed:{}".format(e))
        logging.info("Got metrics: " + msg.str)

    def resetMetrics(self):
        try:
            self.stub.resetMetrics(recall_pb2.ServerMessage())
        except Exception as e:
            logging.warning("RPC failed:{}".format(e))


def single_thread_client(client, ids, k, timer):
    result_dict = dict()
    for id in ids:
        search_start = time.perf_counter()
        results = client.search(id, k)
        result_dict[id] = results
        print(id, ":", results)
        search_end = time.perf_counter()
        timer.add(search_end - search_start)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Friesian')
    parser.add_argument('--data_dir', type=str, default="item_ebd.parquet",
                        help='path of parquet files')
    parser.add_argument('--target', type=str, default=None,
                        help='The target of recall service url')
    logging.basicConfig(filename="client.log", level=logging.INFO)
    args = parser.parse_args()

    df = pd.read_parquet(args.data_dir)
    id_list = df["tweet_id"].unique()
    n_thread = 4

    with grpc.insecure_channel(args.target) as channel:
        stub = recall_pb2_grpc.RecallStub(channel)
        client = SimilarityClient(stub)

        client_timer = Timer()
        thread_list = []
        size = len(id_list) // n_thread
        for ith in range(n_thread):
            ids = id_list[ith * size: (ith + 1) * size]
            thread = threading.Thread(target=single_thread_client,
                                      args=(client, ids, 10, client_timer))
            thread_list.append(thread)
            thread.start()

        for thread in thread_list:
            thread.join()

        count, avg = client_timer.get_stat()
        client_timer.reset()
        client.getMetrics()
        client.resetMetrics()
